Design proposal: secure registry support

GeekArthur opened the pull request:
Signed-off-by: jingfu wang <jingfu.j.wang@ibm.com>

**What type of PR is this?**
/kind documentation
[skip ci]

**Which issue(s) this PR fixes**:
Parent issue: https://github.com/openshift/odo/issues/2893

GeekArthur commented:
/assign

GeekArthur commented:
> Looks good. It looks like you have addressed my early comments on the design from our previous conversation already. One thing that I think is missing is slightly beyond the catalog list, when someone is doing `odo create` on a component that belongs to a secure registry, we'll need to make sure we handle the component creation properly (including download of the devfile from the secure repo) using same secure setting.

Thanks for reviewing! I add the `Create devfile component from secure registry` section to the design proposal.

kadel commented on docs/proposals/secure-registry.md:
```diff
@@ -0,0 +1,75 @@
+1. Collect Credential
+
+    Currently we have `odo registry add <registry name> <registry URL>` and `odo registry update <registry name> <registry URL>` to add and update registry accordingly. Regarding the CLI design, we can implement the following CLI design to support collecting credential:
+    - `odo registry add <registry name> <registry URL> --token <token>`
```
What would be a command for adding registry with basic-auth?

I can see two options:

`odo registry add <registry name> <registry URL> --user <user> --password <password>`
or 
`odo registry add <registry name> <registry URL> --basic-auth <user>:<password>`

GeekArthur replied:
I like the first option as it aligns with `kubectl`, `docker`, `oc` and `odo`, I am gonna update the design proposal.

GeekArthur commented:
@elsony @kadel Comments addressed, please review again, thanks

gorkem commented:
Also consider https://github.com/devfile/kubernetes-api/issues/82

GeekArthur commented:
> Also consider [devfile/kubernetes-api#82](https://github.com/devfile/kubernetes-api/issues/82). I would hold a decision until it is concluded on the devfile side. If we decide on the OCI based registry it is possible that some (all?) of the concerns on this document will be addressed.

@gorkem Thanks for sharing the OCI-based registry architecture design, this design proposal only focuses on secure registry support, more specifically it only addresses concerns of authentication and authorization stages for existing supported registry and it's not related with the registry design. We can open separate issue and create another design proposal for the registry design to adopt the architecture of OCI-based registry.

adisky commented on docs/proposals/secure-registry.md:
```diff
@@ -0,0 +1,77 @@
+
+Create devfile component from secure registry:
+
+When downloading devfile from secure registry, we validate if the credential is valid by adding token to the request header and checking the response.
```
we are not doing any validation on `odo registry add <registry name> <registry URL> --user <user> --password <password>`? Just thinking out here, shouldn't we validate credentials and url here? than doing at the time of `odo create`?

GeekArthur replied:
Regarding arguments/flags validation, the design is that we divide the validation into two stages:
Early validation stage: Only validate URL basic format, happens when user adds registry URL
Validation stage: Validate URL accessibility, happens when user uses registry URL

That design is based on `yum-config-manager`, design proposal with details under this thread: https://github.com/openshift/odo/pull/2940#discussion_r417440866

kadel commented:
/approve

elsony commented:
/lgtm
